internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,161 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	next wrpkit.Handler
+	// queue for wrp messages
+	queue PriorityQueue
+
+	m sync.Mutex
+	// ingestQueue blocks if the queue ingestion is already in progress, i.e.: send
+	// as many queued messages as possible
+	ingestQueue chan bool
+	// shutdown shuts down the queue ingestion
+	shutdown context.CancelFunc
+}
+
+// New creates a new instance of the Handler struct.  The parameter next is the
+// handler that will be called and monitored for errors.
+func New(next websocket.Egress, opts ...Option) (*Handler, error) {
+	if next == nil {
+		return nil, ErrInvalidInput
+	}
+
+	h := &Handler{
+		next:        next,
+		ingestQueue: make(chan bool, 1),
+	}
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(h); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return h, nil
+}
+
+// Start starts the qos queue ingestion. Called during a ws connect event.
+func (h *Handler) Start() {
+	h.m.Lock()
+	defer h.m.Unlock()
+
+	if h.shutdown != nil {
+		return
+	}
+
+	// reset
+	var ctx context.Context
+	ctx, h.shutdown = context.WithCancel(context.Background())
+
+	go h.run(ctx)
+
+	select {
+	case h.ingestQueue <- true:
+		// send as many queued messages as possible
+	default:
+		// queue ingestion is already in progress
+	}
+
+}
+
+// Stop stops the qos queue ingestion. Called during a ws disconnect event.
+func (h *Handler) Stop() {
+	h.m.Lock()
+	shutdown := h.shutdown
+	// allows qos to restart
+	h.shutdown = nil
+	h.m.Unlock()
+
+	if shutdown == nil {
+		return
+	}
+
+	shutdown()
+}
+
+// HandleWrp is called to queue a message and then attempt to send as many queued messages as possible.
+func (h *Handler) HandleWrp(msg wrp.Message) error {
+	// queue newest message
+	h.queue.Enqueue(msg)
+	select {
+	case h.ingestQueue <- true:
+		// send as many queued messages as possible
+	default:
+		// queue ingestion is already in progress
+	}
+
+	return nil
+}
+
+// run sends as many queued messages as possible until the following occurs:
+// 1. the qos context was cancelled
+// 2. qos queue is empty
+// 3. there was a delivery failure
+// Note, Start() will automatically trigger a SendQueuedMessages().
+func (h *Handler) run(ctx context.Context) {
```

Comment:
You can simplify this code dramatically by just spawning (1) longrunning goroutine.  The complexity of managing the goroutine's lifecycle is definitely codesmell that indicates the code is working too hard.

You basically have (3) signals to manage:
- a done channel, so that you can control the goroutine's shutdown
- a message channel
- a "connection ready" channel for the websocket

You'll need a `Connection` type with a method that returns a signaling channel that indicates the IO is done.  `gorilla/mux` and `gobwas` have this, but not sure if the ws library you're using has it.  You can roll your own with something like:

```
type Connection struct {
    l sync.Mutex
    w io.Writer
}

func (c *Connection) Write(m wrp.Message) <-chan struct{} {
    // will probably want to protect this with a counting semaphore
    // and/or just have a longrunning goroutine that services the connection
    ready := make(chan struct{})
    go func() {
        defer close(ready)

        c.l.Lock()
        defer c.l.Unlock()
        c.w.Write(/*whatever*/)
    }()

    return ready
}
```

Pseudocode:

```
func serviceQOS(done <-chan struct{}, msg <- chan wrp.Message, conn *Connection) {
    // create and manage the priority queue here:
    var pq PriorityQueue

    // signaling channel from the connection
    var ready <-chan struct{}

    for {
        select {
            case <- done:
                return

            case m := <-msg:
                pq.Push(m)

             case <-ready:
                 ready = nil
                 if top, ok := pq.Pop(); ok {
                     ready := conn.Write(top)
                 }
        }
    }
}
```

With the above, you're basically leverage the go runtime to do most of the things this PR is doing.